### PR TITLE
add doc for savepoint

### PR DESCRIPTION
**Pull request by crazycs520**
Signed-off-by: crazycs520 <crazycs520@gmail.com>

### First-time contributors' checklist 

- [ ] I've signed [**Contributor License Agreement**](https://cla-assistant.io/pingcap/docs-cn) that's required for repo owners to accept my contribution.

### What is changed, added or deleted? (Required)

add doc for savepoint.

mostly reference https://dev.mysql.com/doc/refman/8.0/en/savepoint.html

### Which TiDB version(s) do your changes apply to? (Required)

**Tips for choosing the affected version(s):**

By default, **CHOOSE MASTER ONLY** so your changes will be applied to the next TiDB major or minor releases. If your PR involves a product feature behavior change or a compatibility change, **CHOOSE THE AFFECTED RELEASE BRANCH(ES) AND MASTER**.

For details, see [tips for choosing the affected versions (in Chinese)](https://github.com/pingcap/docs-cn/blob/master/CONTRIBUTING.md#版本选择指南).

- [x] master (the latest development version)
- [x] v6.2 (TiDB 6.2 versions)
- [ ] v6.1 (TiDB 6.1 versions)
- [ ] v6.0 (TiDB 6.0 versions)
- [ ] v5.4 (TiDB 5.4 versions)
- [ ] v5.3 (TiDB 5.3 versions)
- [ ] v5.2 (TiDB 5.2 versions)
- [ ] v5.1 (TiDB 5.1 versions)
- [ ] v5.0 (TiDB 5.0 versions)
- [ ] v4.0 (TiDB 4.0 versions)

### What is the related PR or file link(s)?

- Other reference link(s): https://github.com/pingcap/tidb/pull/34466

### Do your changes match any of the following descriptions?

- [ ] Delete files
- [ ] Change aliases
- [ ] Need modification after applied to another branch 
- [ ] Might cause conflicts after applied to another branch

### CODE REVIEW

**Comment by easonn7** [comment]
LGTM

File: develop/dev-guide-transaction-restraints.md [hunk 1]
```diff
@@ -716,7 +716,7 @@ mysql> SELECT * FROM T2;
 +------+
 ```
 
-TiDB 不支持 savepoint 机制，因此也不支持 PROPAGATION_NESTED 传播行为。基于 Java Spring 框架的应用如果使用了 PROPAGATION_NESTED 传播行为，需要在应用端做出调整，将嵌套事务的逻辑移除。
+TiDB v6.2.0 版本才支持 [savepoint](/sql-statements/sql-statement-savepoint.md) 特性。因此低于 v6.2.0 版本的 TiDB 是不支持 PROPAGATION_NESTED 传播行为。基于 Java Spring 框架的应用如果使用了 PROPAGATION_NESTED 传播行为，需要在应用端做出调整，将嵌套事务的逻辑移除。
```
Comment (Icemap) [inline develop/dev-guide-transaction-restraints.md:719]:
```suggestion
TiDB v6.2.0 版本开始支持 [savepoint](/sql-statements/sql-statement-savepoint.md) 特性。因此低于 v6.2.0 版本的 TiDB 是不支持 PROPAGATION_NESTED 传播行为的。基于 Java Spring 框架的应用如果使用了 PROPAGATION_NESTED 传播行为，需要在应用端做出调整，将嵌套事务的逻辑移除。
```

File: sql-statements/sql-statement-savepoint.md [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+`SAVEPOINT` 是 TiDB 从 v6.2.0 开始支持的特性，语法如下：
+
+```sql
+SAVEPOINT identifier
+ROLLBACK TO [SAVEPOINT] identifier
+RELEASE SAVEPOINT identifier
+```
+
+`SAVEPOINT` 语句用于在当前事务中，设置一个指定名字保存点。如果已经存在相同名字的保存点，就删除旧的保存点后设置新的保存点。
```
Comment (Icemap) [inline sql-statements/sql-statement-savepoint.md:17]:
```suggestion
`SAVEPOINT` 语句用于在当前事务中，设置一个指定名称的保存点。如果已经存在该名称的保存点，将会更新保存点到当前位置。
```

File: sql-statements/sql-statement-savepoint.md [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+`ROLLBACK TO SAVEPOINT` 语句将事务回滚到指定名字的事务保存点，而不终止该事务。当前事务在设置保存点后，对表数据所做的修改将在回滚中撤销，且删除事务保存点之后的所有保存点。悲观事务中，对于已经持有的悲观锁不会回滚，而是在事务结束时才释放。
```
Comment (Icemap) [inline sql-statements/sql-statement-savepoint.md:19]:
```suggestion
`ROLLBACK TO SAVEPOINT` 语句将事务回滚到指定名称的事务保存点，而不终止该事务。当前事务在设置保存点后，对表数据所做的修改将在回滚中撤销，且删除事务保存点之后的所有保存点。悲观事务中，对于已经持有的悲观锁不会回滚，而是在事务结束时才释放。
```

File: sql-statements/sql-statement-savepoint.md [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+
+如果 `ROLLBACK TO SAVEPOINT` 语句中指定名字的保存点不存在，则会返回以下报错：
```
Comment (Icemap) [inline sql-statements/sql-statement-savepoint.md:21]:
```suggestion
如果 `ROLLBACK TO SAVEPOINT` 语句中指定名称的保存点不存在，则会返回以下错误信息：
```

File: sql-statements/sql-statement-savepoint.md [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+```
+ERROR 1305 (42000): SAVEPOINT identifier does not exist
+```
+
+`RELEASE SAVEPOINT` 语句从当前事务中删除指定名字的保存点以及该保存点之后的所有保存点。不会提交或回滚当前事务。如果指定的保存点不存在，则会返回以下报错：
```
Comment (Icemap) [inline sql-statements/sql-statement-savepoint.md:27]:
```suggestion
`RELEASE SAVEPOINT` 语句将从当前事务中删除指定名称及之后的**_所有_**保存点。不会提交或回滚当前事务。如果指定名称的保存点不存在，则会返回以下错误信息：
```

File: sql-statements/sql-statement-savepoint.md [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+```
+
+当事务提交或者回滚后，事务中所有保存点都会被删除。
+
+## 示例
+
+{{< copyable "sql" >}}
+
+```sql
+CREATE TABLE t1 (a int NOT NULL PRIMARY KEY);
```
Comment (Icemap) [inline sql-statements/sql-statement-savepoint.md:40]:
```suggestion
CREATE TABLE `t1` (`a` INT NOT NULL PRIMARY KEY);
```

File: sql-statements/sql-statement-savepoint.md [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+```sql
+SAVEPOINT sp1;
+```
+
+```
+Query OK, 0 rows affected (0.01 sec)
+```
+
+```sql
+INSERT INTO t1 VALUES (2);
```
Comment (Icemap) [inline sql-statements/sql-statement-savepoint.md:78]:
```suggestion
INSERT INTO `t1` VALUES (2);
```

File: sql-statements/sql-statement-savepoint.md [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+```
+
+```
+Query OK, 0 rows affected (0.01 sec)
+```
+
+{{< copyable "sql" >}}
+
+```sql
+SELECT * FROM t1;
```
Comment (Icemap) [inline sql-statements/sql-statement-savepoint.md:128]:
```suggestion
SELECT * FROM `t1`;
```

File: sql-statements/sql-statement-savepoint.md [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+```
+
```
Comment (Icemap) [inline sql-statements/sql-statement-savepoint.md:130]:
```suggestion

运行结果为：

```

File: sql-statements/sql-statement-savepoint.md [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+```
+Query OK, 0 rows affected (0.01 sec)
+```
+
+{{< copyable "sql" >}}
+
+```sql
+COMMIT;
+```
+
```
Comment (Icemap) [inline sql-statements/sql-statement-savepoint.md:120]:
```suggestion

运行结果为：

```

File: sql-statements/sql-statement-savepoint.md [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+```
+Query OK, 0 rows affected (0.01 sec)
+```
+
+{{< copyable "sql" >}}
+
+```sql
+ROLLBACK TO SAVEPOINT sp1;
+```
+
```
Comment (Icemap) [inline sql-statements/sql-statement-savepoint.md:110]:
```suggestion

运行结果为：

```

File: sql-statements/sql-statement-savepoint.md [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+```
+Query OK, 0 rows affected (0.01 sec)
+```
+
+{{< copyable "sql" >}}
+
+```sql
+RELEASE SAVEPOINT sp2;
+```
+
```
Comment (Icemap) [inline sql-statements/sql-statement-savepoint.md:100]:
```suggestion

运行结果为：

```

File: sql-statements/sql-statement-savepoint.md [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+```
+Query OK, 1 row affected (0.00 sec)
+```
+
+{{< copyable "sql" >}}
+
+```sql
+SAVEPOINT sp2;
+```
+
```
Comment (Icemap) [inline sql-statements/sql-statement-savepoint.md:90]:
```suggestion

运行结果为：

```

File: sql-statements/sql-statement-savepoint.md [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+```
+Query OK, 0 rows affected (0.01 sec)
+```
+
+```sql
+INSERT INTO t1 VALUES (2);
+```
+
```
Comment (Icemap) [inline sql-statements/sql-statement-savepoint.md:80]:
```suggestion

运行结果为：

```

File: sql-statements/sql-statement-savepoint.md [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+```
+Query OK, 1 row affected (0.00 sec)
+```
+
+{{< copyable "sql" >}}
+
+```sql
+SAVEPOINT sp1;
+```
+
```
Comment (Icemap) [inline sql-statements/sql-statement-savepoint.md:72]:
```suggestion

运行结果为：

```

File: sql-statements/sql-statement-savepoint.md [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+```
+Query OK, 0 rows affected (0.00 sec)
+```
+
+{{< copyable "sql" >}}
+
+```sql
+INSERT INTO t1 VALUES (1);
+```
+
```
Comment (Icemap) [inline sql-statements/sql-statement-savepoint.md:62]:
```suggestion

运行结果为：

```

File: sql-statements/sql-statement-savepoint.md [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+```
+Query OK, 0 rows affected (0.12 sec)
+```
+
+{{< copyable "sql" >}}
+
+```sql
+BEGIN;
+```
+
```
Comment (Icemap) [inline sql-statements/sql-statement-savepoint.md:52]:
```suggestion

运行结果为：

```

File: sql-statements/sql-statement-savepoint.md [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+当事务提交或者回滚后，事务中所有保存点都会被删除。
+
+## 示例
+
+{{< copyable "sql" >}}
+
+```sql
+CREATE TABLE t1 (a int NOT NULL PRIMARY KEY);
+```
+
```
Comment (Icemap) [inline sql-statements/sql-statement-savepoint.md:42]:
```suggestion

运行结果为：

```

File: develop/dev-guide-choose-driver-or-orm.md [hunk 19]
```diff
@@ -52,25 +52,8 @@ TiDB 与 MySQL 有差异：
 
 > **注意：**
 >
-> Hibernate 当前[不支持嵌套事务](https://stackoverflow.com/questions/37927208/nested-transaction-in-spring-app-with-jpa-postgres)，TiDB 当前版本也[不支持 Savepoint](https://github.com/pingcap/tidb/issues/6840)。
+> Hibernate 当前[不支持嵌套事务](https://stackoverflow.com/questions/37927208/nested-transaction-in-spring-app-with-jpa-postgres)，TiDB v6.2.0 版本才开始支持[Savepoint](/sql-statements/sql-statement-savepoint.md)。
```
Comment (bb7133) [inline develop/dev-guide-choose-driver-or-orm.md:55]:
```suggestion
> Hibernate 当前[不支持嵌套事务](https://stackoverflow.com/questions/37927208/nested-transaction-in-spring-app-with-jpa-postgres)，TiDB v6.2.0 版本开始支持[Savepoint](/sql-statements/sql-statement-savepoint.md)。
```

File: sql-statements/sql-statement-savepoint.md (hunk 2 above)
Comment (crazycs520) [reply sql-statements/sql-statement-savepoint.md:17]: 不是这个意思，不是更新。

File: sql-statements/sql-statement-savepoint.md (hunk 17 above)
Comment (crazycs520) [reply sql-statements/sql-statement-savepoint.md:52]: 不用插这个吧，感觉有点点啰嗦...

**Review by bb7133 — APPROVED** [review]
LGTM

**Comment by crazycs520** [comment]
@TomShawn PTAL

**Comment by crazycs520** [comment]
@TomShawn PTAL

**Comment by shichun-0415** [comment]
> @TomShawn PTAL

@crazycs520 As is confirmed, we will merge this PR upon v6.2 release. For now, we will hold it. cc @TomShawn

File: sql-statements/sql-statement-savepoint.md [hunk 20]
```diff
@@ -0,0 +1,152 @@
...
+| a |
++---+
+| 1 |
++---+
+1 row in set
+```
+
+## MySQL 兼容性
+
+`ROLLBACK TO SAVEPOINT` 语句在回滚到指定保存点时，MySQL 会释放保存点之后才持有的锁，但在 TiDB 悲观事务中，不会立即释放保存点之后才持有的锁，而是会等到事务提交或者回滚时，才释放所有持有的锁。
```
Comment (TomShawn) [inline sql-statements/sql-statement-savepoint.md:142]:
```suggestion
使用 `ROLLBACK TO SAVEPOINT` 语句将事物回滚到指定保存点时，MySQL 会释放该保存点之后才持有的锁，但在 TiDB 悲观事务中，不会立即释放该保存点之后才持有的锁，而是等到事务提交或者回滚时，才释放全部持有的锁。
```

File: sql-statements/sql-statement-savepoint.md [hunk 21]
```diff
@@ -0,0 +1,152 @@
...
+
+`SAVEPOINT` 特性不支持与 TiDB Binlog 一起使用。
```
Comment (TomShawn) [inline sql-statements/sql-statement-savepoint.md:144]: 这一句能否放在最前面，例如 L9。

**Review by TomShawn — COMMENTED** [review]
- Refine sentences to make them more readable
- Add descriptions for the SQL statements to make the information more user-friendly
- Request clarification for a place
- Suggest the change of information sequence to highlight compatibility issue

File: sql-statements/sql-statement-savepoint.md [hunk 22]
```diff
@@ -0,0 +1,152 @@
...
+ERROR 1305 (42000): SAVEPOINT identifier does not exist
+```
+
+`RELEASE SAVEPOINT` 语句将从当前事务中删除指定名称及之后的**_所有_**保存点。不会提交或回滚当前事务。如果指定名称的保存点不存在，则会返回以下错误信息：
+
+```
+ERROR 1305 (42000): SAVEPOINT identifier does not exist
+```
+
+当事务提交或者回滚后，事务中所有保存点都会被删除。
```
[the remark this answers is not in the document]
Comment (crazycs520) [reply sql-statements/sql-statement-savepoint.md:32]: 这里不用指定名称，事务在提交或回滚后，表示事务已经结束了，所以这个事务中的所有保存点都会被删除。

File: sql-statements/sql-statement-savepoint.md [hunk 23]
```diff
@@ -0,0 +1,165 @@
...
+| 1 |
++---+
+1 row in set
+```
+
+## MySQL 兼容性
+
+使用 `ROLLBACK TO SAVEPOINT` 语句将事物回滚到指定保存点时，MySQL 会释放该保存点之后才持有的锁，但在 TiDB 悲观事务中，不会立即释放该保存点之后才持有的锁，而是等到事务提交或者回滚时，才释放全部持有的锁。
+
+`SAVEPOINT` 特性不支持与 TiDB Binlog 一起使用。
```
Comment (crazycs520) [inline sql-statements/sql-statement-savepoint.md:157]:
```suggestion
`SAVEPOINT` 特性不支持与 TiDB Binlog 一起使用。

使用 `ROLLBACK TO SAVEPOINT` 语句将事物回滚到指定保存点时，MySQL 会释放该保存点之后才持有的锁，但在 TiDB 悲观事务中，不会立即释放该保存点之后才持有的锁，而是等到事务提交或者回滚时，才释放全部持有的锁。
```

File: sql-statements/sql-statement-savepoint.md (hunk 21 above)
Comment (crazycs520) [reply sql-statements/sql-statement-savepoint.md:155]: 不用吧，放在这个兼容性这里比较合适

File: sql-statements/sql-statement-savepoint.md [hunk 24]
```diff
@@ -0,0 +1,152 @@
+---
+title: SAVEPOINT
+summary: TiDB 数据库中 SAVEPOINT 的使用概况。
+aliases: ['/docs-cn/dev/sql-statements/sql-statement-savepoint/','/docs-cn/dev/reference/sql/statements/savepoint/']
+---
+
+# SAVEPOINT
+
+`SAVEPOINT` 是 TiDB 从 v6.2.0 开始支持的特性，语法如下：
```
[the remark this answers is not in the document]
Comment (crazycs520) [reply sql-statements/sql-statement-savepoint.md:8]: https://dev.mysql.com/doc/refman/8.0/en/savepoint.html

mysql 的官方文档里面也没说使用场景的，只有功能介绍

**Review by TomShawn — APPROVED** [review]
LGTM

File: develop/dev-guide-choose-driver-or-orm.md [hunk 25]
```diff
@@ -52,25 +52,8 @@ TiDB 与 MySQL 有差异：
 
 > **注意：**
 >
-> Hibernate 当前[不支持嵌套事务](https://stackoverflow.com/questions/37927208/nested-transaction-in-spring-app-with-jpa-postgres)，TiDB 当前版本也[不支持 Savepoint](https://github.com/pingcap/tidb/issues/6840)。
+> Hibernate 当前[不支持嵌套事务](https://stackoverflow.com/questions/37927208/nested-transaction-in-spring-app-with-jpa-postgres)，TiDB 从 v6.2.0 版本开始支持[Savepoint](/sql-statements/sql-statement-savepoint.md)。
 > 若你使用 Spring Data JPA 等框架，在 **@Transactional** 中请勿使用 `Propagation.NESTED` 事务传播选项，即：`@Transactional(propagation = Propagation.NESTED)`。
```
Comment (TomShawn) [inline develop/dev-guide-choose-driver-or-orm.md:56]:
```suggestion
> Hibernate 当前[不支持嵌套事务](https://stackoverflow.com/questions/37927208/nested-transaction-in-spring-app-with-jpa-postgres)，TiDB 从 v6.2.0 版本开始支持[Savepoint](/sql-statements/sql-statement-savepoint.md)。
>
> 若你使用 Spring Data JPA 等框架，在 **@Transactional** 中请勿使用 `Propagation.NESTED` 事务传播选项，即：`@Transactional(propagation = Propagation.NESTED)`。
```

File: develop/dev-guide-choose-driver-or-orm.md [hunk 26]
```diff
@@ -52,25 +52,9 @@ TiDB 与 MySQL 有差异：
...
-> Hibernate 当前[不支持嵌套事务](https://stackoverflow.com/questions/37927208/nested-transaction-in-spring-app-with-jpa-postgres)，TiDB 当前版本也[不支持 Savepoint](https://github.com/pingcap/tidb/issues/6840)。
-> 若你使用 Spring Data JPA 等框架，在 **@Transactional** 中请勿使用 `Propagation.NESTED` 事务传播选项，即：`@Transactional(propagation = Propagation.NESTED)`。
->
-> 你可以使用[这个例子](https://github.com/Icemap/tidb-savepoint)，快速复现 TiDB 与 MySQL 对 Savepoint 的输出结果：
->
-> ```
-> MySQL:
-> id: 1, coins: 1, goods: 1
-> id: 3, coins: 1, goods: 1
+> Hibernate 当前[不支持嵌套事务](https://stackoverflow.com/questions/37927208/nested-transaction-in-spring-app-with-jpa-postgres)，TiDB 从 v6.2.0 版本开始支持[Savepoint](/sql-statements/sql-statement-savepoint.md)。
```
Comment (TomShawn) [inline develop/dev-guide-choose-driver-or-orm.md:55]:
```suggestion
> Hibernate 当前[不支持嵌套事务](https://stackoverflow.com/questions/37927208/nested-transaction-in-spring-app-with-jpa-postgres)，TiDB 从 v6.2.0 版本开始支持 [Savepoint](/sql-statements/sql-statement-savepoint.md)。
```

File: sql-statements/sql-statement-savepoint.md [hunk 27]
```diff
@@ -0,0 +1,149 @@
...
+
+```sql
+SAVEPOINT identifier
+ROLLBACK TO [SAVEPOINT] identifier
+RELEASE SAVEPOINT identifier
+```
+
+> **警告：**
+>
```
Comment (Oreoxmt) [inline sql-statements/sql-statement-savepoint.md:17]:
```suggestion
>
```

File: best-practices/java-app-best-practices.md [hunk 28]
```diff
@@ -301,7 +301,7 @@ Cursor<Post> queryAllPost();
 
 在应用代码中业务可能会通过使用 [Spring Transaction](https://docs.spring.io/spring/docs/4.2.x/spring-framework-reference/html/transaction.html) 和 AOP 切面的方式来启停事务。
 
-通过在方法定义上添加 `@Transactional` 注解标记方法，AOP 将会在方法前开启事务，方法返回结果前 commit 事务。如果遇到类似业务，可以通过查找代码 `@Transactional` 来确定事务的开启和关闭时机。需要特别注意有内嵌的情况，如果发生内嵌，Spring 会根据 [Propagation](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/transaction/annotation/Propagation.html) 配置使用不同的行为，因为 TiDB 未支持 savepoint，所以不支持嵌套事务。
+通过在方法定义上添加 `@Transactional` 注解标记方法，AOP 将会在方法前开启事务，方法返回结果前 commit 事务。如果遇到类似业务，可以通过查找代码 `@Transactional` 来确定事务的开启和关闭时机。需要特别注意有内嵌的情况，如果发生内嵌，Spring 会根据 [Propagation](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/transaction/annotation/Propagation.html) 配置使用不同的行为。
```
Comment (Oreoxmt) [inline best-practices/java-app-best-practices.md:304]:
```suggestion
通过在方法定义上添加 `@Transactional` 注解标记方法，AOP 将会在方法前开启事务，方法返回结果前提交事务。如果遇到类似业务，可以通过查找代码 `@Transactional` 来确定事务的开启和关闭时机。需要特别注意有内嵌的情况，如果发生内嵌，Spring 会根据 [Propagation](https://docs.spring.io/spring-framework/docs/current/javadoc-api/org/springframework/transaction/annotation/Propagation.html) 配置使用不同的行为。
```

File: sql-statements/sql-statement-savepoint.md [hunk 29]
```diff
@@ -0,0 +1,149 @@
...
+
+```sql
+ROLLBACK TO SAVEPOINT sp1;
+```
+
+```sql
+Query OK, 0 rows affected (0.01 sec)
+```
+
+提交事务并查询表格，发现表中仅有 `sp1` 前插入的数据。
```
Comment (Oreoxmt) [inline sql-statements/sql-statement-savepoint.md:116]:
```suggestion
提交事务并查询表格，发现表中仅有 `sp1` 前插入的数据：
```

**Review by Oreoxmt — APPROVED** [review]
Rest LGTM

**Pull request merged**